src/connectionInfo/connectionManager.ts

```diff
@@ -0,0 +1,159 @@
+import { isIPv4 } from 'net';
+import * as vscode from 'vscode';
+import * as nls from 'vscode-nls';
+import { DEFAULT_HOST } from '../utils/constants';
+import { Disposable } from '../utils/dispose';
+import { SETTINGS_SECTION_ID, SettingUtil } from '../utils/settingsUtil';
+import { Connection, ConnectionInfo } from './connection';
+
+/**
+ * @description keeps track of all of the Connection objects and the info needed to create them (ie: initial ports).
+ */
+const localize = nls.loadMessageBundle();
+export class ConnectionManager extends Disposable {
+	private _initHttpPort;
+	private _initWSPort;
+	private _initHost: string;
+	private _connections: Map<vscode.Uri | undefined, Connection>; // undefined key means no workspace root
```

Comment:
Safest thing to do is to use a uri's toString() as a map key instead of a URI object. If you have two URIs that have the same value, but are different objects, then they will be different keys in the map.